geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java

```diff
@@ -0,0 +1,146 @@
+    assertThat(jedis.lindex(NON_EXISTENT_LIST_KEY, -2)).isNull();
+  }
+
+  @Test
+  public void lindex_withPositiveIndexes_returnsElement() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+    for (int i = 0; i < LIST_ELEMENTS.length; i++) {
+      // LIST_ELEMENTS[LIST_ELEMENTS.length - 1 - i] iterates LIST_ELEMENTS backwards
+      System.out
+          .println(jedis.lindex(LIST_KEY, i) + " " + LIST_ELEMENTS[LIST_ELEMENTS.length - 1 - i]);
+      assertThat(jedis.lindex(LIST_KEY, i)).isEqualTo(LIST_ELEMENTS[LIST_ELEMENTS.length - 1 - i]);
+    }
+  }
+
+  @Test
+  public void lindex_withNegativeIndexes_returnsElement() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+    for (int i = -7; i < 0; i++) {
+      // LIST_ELEMENTS[-(i + 1)] iterates LIST_ELEMENTS forwards
+      assertThat(jedis.lindex(LIST_KEY, i)).isEqualTo(LIST_ELEMENTS[-(i + 1)]);
+    }
+  }
+
+  @Test
+  public void lindex_withPositiveOutOfRangeIndex_returnsNull() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+    assertThat(jedis.lindex(LIST_KEY, 10)).isNull();
+  }
+
+  @Test
+  public void lindex_withNegativeOutOfRangeIndex_returnsNull() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+    assertThat(jedis.lindex(LIST_KEY, -10)).isNull();
+  }
+
+  @Test
+  public void lindex_withInvalidIndex_withNonExistentList_returnsNull() {
+    assertThat(jedis.sendCommand(NON_EXISTENT_LIST_KEY, Protocol.Command.LINDEX,
+        NON_EXISTENT_LIST_KEY, "b")).isNull();
+  }
+
+  @Test
+  public void lindex_withInvalidIndex_returnsError() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+    assertThatThrownBy(() -> jedis.sendCommand(LIST_KEY, Protocol.Command.LINDEX, LIST_KEY, "b"))
+        .hasMessage("ERR " + ERROR_NOT_INTEGER);
+  }
+
+  @Test
+  public void lindex_withWrongKeyType_returnsWrongTypeError() {
+    String key = "{tag1}ding";
+    jedis.set(key, "dong");
+    assertThatThrownBy(() -> jedis.lindex(key, 2)).hasMessage("WRONGTYPE " + ERROR_WRONG_TYPE);
+  }
+
+  @Test
+  public void lindex_withWrongKeyType_withInvalidIndex_returnsWrongTypeError() {
+    String key = "{tag1}ding";
+    jedis.set(key, "dong");
+    assertThatThrownBy(() -> jedis.sendCommand(LIST_KEY, Protocol.Command.LINDEX, key, "b"))
+        .hasMessage("WRONGTYPE " + ERROR_WRONG_TYPE);
+  }
+
+  @Test
+  public void ensureListConsistency_whenRunningConcurrently() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+
+    String elementToAdd = "zebra";
+    final AtomicReference<String> lindexResultReference = new AtomicReference<>();
+    new ConcurrentLoopingThreads(1000,
+        i -> jedis.lpush(LIST_KEY, elementToAdd),
+        i -> lindexResultReference.set(jedis.lindex(LIST_KEY, 0)))
+            .runWithAction(() -> {
+              assertThat(lindexResultReference).satisfiesAnyOf(
+                  lindexResult -> assertThat(lindexResult.get()).isEqualTo(elementToAdd),
+                  lindexResult -> assertThat(lindexResult.get()).isEqualTo("goat"));
+              jedis.lpop(LIST_KEY);
```

Comment:
Good call on checking index 0 rather than the last index, my brain was thinking about things backwards.

However, it's not possible to spot concurrency issues when only pushing one element (the element is either there or it's not, so we can't ever tell if we tried to get the element at index 0 while elements were still being added), so for this test, a better approach would be to add multiple new elements, check index 0, then delete the whole key and re-add the initial elements at the end of the `runWithAction()` to get us back to the starting conditions.